SourceCode/h8/light.cpp

```diff
@@ -0,0 +1,796 @@
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <cmath>
+#include <iterator>
+#include <vector>
+
+#define MAX_MATRIX_ROWS 4
+#define MAX_MATRIX_COLUMNS 4
+
+class Matrix {
+    float fields[MAX_MATRIX_ROWS][MAX_MATRIX_COLUMNS] = {0};
+    unsigned int rows, columns;
+public:
+    Matrix(const float * const matrix, const unsigned int rows, const unsigned int columns) {
+        if(!matrix) return;
+        this->rows = rows;
+        this->columns = columns;
+        for(unsigned int i = 0; i < rows; i++) {
+            for(unsigned int j = 0; j < columns; j++) {
+                fields[i][j] = matrix[j + columns * i];
+            }
+        }
+    }
+    Matrix(const unsigned int rows, const unsigned int columns) {
+        this->rows = rows;
+        this->columns = columns;
+    }
+
+    void scale(const float num) {
+        for(unsigned int i = 0; i < rows; i++) {
+            for(unsigned int j = 0; j < columns; j++) {
+                fields[i][j] *= num;
+            }
+        }
+    }
+    void bound(const float num) {
+        for(unsigned int i = 0; i < rows; i++) {
+            for(unsigned int j = 0; j < columns; j++) {
+                if(fields[i][j] > num) fields[i][j] = num;
+            }
+        }
+    }
+    void transpose() {
+        for(unsigned int i = 0; i < rows-1; i++) {
+            for(unsigned int j = i+1; j < columns; j++) {
+                const float temp = fields[i][j];
+                fields[i][j] = fields[j][i];
+                fields[j][i] = temp;
+            }
+        }
+    }
+
+    unsigned int getRows() const {
+        return rows;
+    }
+    unsigned int getColumns() const {
+        return columns;
+    }
+    float getField(const unsigned int i, const unsigned int j) const {
+        return fields[i][j];
+    }
+    float (&getFields())[MAX_MATRIX_ROWS][MAX_MATRIX_COLUMNS] {
+        return fields;
+    }
+
+    void setRows(const unsigned int rows) {
+        this->rows = rows;
+    }
+    void setColumns(const unsigned int columns) {
+        this->columns = columns;
+    }
+    void setField(const unsigned int i, const unsigned int j, const float value) {
+        fields[i][j] = value;
+    }
+    void setFields(const float * const matrix) {
+        for(unsigned int i = 0; i < rows; i++) {
+            for(unsigned int j = 0; j < columns; j++) {
+                fields[i][j] = matrix[j + columns * i];
+            }
+        }
+    }
+
+    void print() const {
+        for(unsigned int i = 0; i < rows; i++) {
+            for(unsigned int j = 0; j < columns; j++) {
+                std::cout << fields[i][j] << " ";
+            }
+            std::cout << std::endl;
+        }
+    }
+};
+
+class Vector : public Matrix {
```

Comment:
This seems wrong, Vector should not inherit from Matrix!

Reply:
Mathematically speaking a Vector is a special case of a matrix (1x3 or 3x1). In addition, it allows me to use matrix operations and multiplications on vector without code duplication. @VladislavVulchevChaos